packages/cozy-harvest-lib/src/components/OAuthWindow.jsx

```diff
@@ -52,29 +53,58 @@ export class OAuthWindow extends PureComponent {
     if (typeof onCancel === 'function') onCancel()
   }
 
-  handleMessage(messageEvent) {
+  /**
+   * Handles OAuth data. OAuth data may be provided by different way:
+   * * postMessage from web apps (see handleMessage)
+   * * url changes from mobile apps
+   * @param  {string} data.key `io.cozy.accounts` id The created OAuth account
+   * @param  {string} data.oAuthStateKey key for localStorage
+   */
+  handleOAuthData(data) {
     const { konnector, onSuccess } = this.props
-
-    const { data } = messageEvent
-
     if (!checkOAuthData(konnector, data)) return
     this.setState({ succeed: true })
 
     if (typeof onSuccess !== 'function') return
+    this.setState({ succeed: true })
     onSuccess(data.key)
   }
 
+  /**
+   * Expects receiving message from web app
+   * @param  {MessageEvent} messageEvent
+   */
+  handleMessage(messageEvent) {
+    this.handleOAuthData(messageEvent.data)
+  }
+
+  /**
+   * Monitor URL changes for mobile apps and in app browsers
+   * @param  {URL} url
+   */
+  handleUrlChange(url) {
+    const account = url.searchParams.get('account')
+    const state = url.searchParams.get('state')
+
+    this.handleOAuthData({
```

Comment:
We should `handleOauthData()` only if we have account and state. Because we'll call this method several times (imagine we have a two step login page : 
- first page : email => then click => redirect to new page => enter password => ...)

Reply:
I think I see your point: `key` and `oAuthStateKey` are already checked here https://github.com/cozy/cozy-libs/pull/462/commits/344e0571af88d8a0b4b3614c38f54fdab2043c2b#diff-0347ea887df5279b52d473835e80382eR18 but I guess it is not explicit enough here ?

Reply:
I really think this is too deep in the process :/